### PR TITLE
Disallow more locals of restricted type in async methods

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/62747

The bug was that we allowed declaring locals for restricted type in async methods inside `foreach`. For example, `foreach (Span<int> s in ...)` in async methods.
Looking around, I noticed a couple other scenarios with the same issue:
1. deconstruction declarations, 
2. `using (expression)` which lowers to use a temp

### CODE REVIEW

**Review by cston — APPROVED** [review]

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs [hunk 1]
```diff
@@ -298,6 +298,11 @@ private BoundForEachStatement BindForEachPartsWorker(BindingDiagnosticBag diagno
                         SourceLocalSymbol local = this.IterationVariable;
                         local.SetTypeWithAnnotations(declType);
 
+                        if (CheckRestrictedTypeInAsyncMethod(this.ContainingMemberOrLambda, declType.Type, diagnostics, typeSyntax))
+                        {
+                            hasErrors = true;
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:303]: >hasErrors = true; [](http://example.com/codeflow?start=28&length=17)

What is our intent here. We don't need to record every possible error in a bound node. The only requirement to set HasError flag is when the node is completely broken. I do not it is the case here. We were able to bind all things.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:303]: Setting HasError without a good reason is likely to disable some useful analysis

File: src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs [hunk 2]
```diff
@@ -116,6 +116,11 @@ internal static BoundStatement BindUsingStatementOrDeclarationFromParts(SyntaxNo
             {
                 expressionOpt = usingBinderOpt!.BindTargetExpression(diagnostics, originalBinder);
                 hasErrors |= !bindDisposable(fromExpression: true, out patternDisposeInfo, out awaitableTypeOpt);
+                Debug.Assert(expressionOpt is not null);
+                if (expressionOpt.Type is not null)
+                {
+                    CheckRestrictedTypeInAsyncMethod(originalBinder.ContainingMemberOrLambda, expressionOpt.Type, diagnostics, expressionOpt.Syntax);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/UsingStatementBinder.cs:122]: >CheckRestrictedTypeInAsyncMethod(originalBinder.ContainingMemberOrLambda, expressionOpt.Type, diagnostics, expressionOpt.Syntax); [](http://example.com/codeflow?start=20&length=129)

Is the wording of the error going to be confusing here?There is no user declared local here

File: src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs [hunk 3]
```diff
@@ -1008,19 +1009,208 @@ public static async Task<int> M1()
...
+            var src = @"
+using System;
+using System.Threading.Tasks;
+
+public class Program
+{
+    public async Task M()
+    {
+        (Span<int> s1, Span<int> s2) = new Program(); // 1, 2
+        var (s3, s4) = new Program(); // 3, 4
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs:1115]: >var (s3, s4) = new Program(); [](http://example.com/codeflow?start=8&length=29)

Consider also testing scenario like ```(var s5, var s6) = new Program();

File: src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs [hunk 4]
```diff
@@ -1008,19 +1009,208 @@ public static async Task<int> M1()
...
+}
+
+public ref struct RS
+{
+    public void Dispose() { }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: TargetFramework.Net70);
+            comp.VerifyDiagnostics(
+                // (8,16): error CS4012: Parameters or locals of type 'RS' cannot be declared in async methods or async lambda expressions.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs:1168]: >// (8,16): error CS4012: Parameters or locals of type 'RS' cannot be declared in async methods or async lambda expressions. [](http://example.com/codeflow?start=16&length=123)

We probably should add a dedicated error for his case, that doesn't talk about locals and parameters

File: src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs [hunk 5]
```diff
@@ -1008,19 +1009,208 @@ public static async Task<int> M1()
...
+using System.Threading.Tasks;
+
+public class Program
+{
+    public async Task M()
+    {
+        using (default(RS)) { } // 1
+        using (var s1 = default(RS)) { } // 2
+        using (RS s2 = default(RS)) { } // 3
+        using RS s3 = default(RS); // 4
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs:1154]: >using RS s3 = default(RS); [](http://example.com/codeflow?start=8&length=26)

Consider testing `var` case as well

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs (hunk 1 above)
Comment (jcouv) [reply src/Compilers/CSharp/Portable/Binder/ForEachLoopBinder.cs:303]: This was out of consistency with other callers of `CheckRestrictedTypeInAsyncMethod`. But I agree this error doesn't prevent us from binding.

**Review by cston — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs [hunk 6]
```diff
@@ -1008,19 +1009,231 @@ public static async Task<int> M1()
...
+}
+
+public ref struct RS
+{
+    public void Dispose() { }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: TargetFramework.Net70);
+            comp.VerifyDiagnostics(
+                // (8,16): error CS9104: Values of type 'RS' cannot be used in async methods or async lambda expressions.
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs:1188]: >Values of type 'RS' cannot be used in async methods or async lambda expressions. [](http://example.com/codeflow?start=41&length=80)

I think that this wording is also confusing, and quite possibly plain incorrect. Values of RS can be used in an async method. For example, the following code compiles today, and I think will continue compiling with changes in this PR, but it is using value of `RS` type in an async method:
[https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAIAYAEGCMArANwCwAUBRgMx4BMe+A7BQN4U6d60YAceANhwBZABQBKDl3bkucvAE5BAOgAiAUxQQAnqPziys+ZwDi6hACUAyhLVQAzgAcA9vfUTDxzhiaeuAXylOIMYhaxwzSxtxHABeAD4cABN1ADMIAFcUBENAynIaHAAnNJx7BCKMgGMEHGs2EMKMABYcVQcXNwkcVhw8/yA===](https://sharplab.io/#v2:EYLgtghglgdgNAFxFANgHwAIAYAEGCMArANwCwAUBRgMx4BMe+A7BQN4U6d60YAceANhwBZABQBKDl3bkucvAE5BAOgAiAUxQQAnqPziys+ZwDi6hACUAyhLVQAzgAcA9vfUTDxzhiaeuAXylOIMYhaxwzSxtxHABeAD4cABN1ADMIAFcUBENAynIaHAAnNJx7BCKMgGMEHGs2EMKMABYcVQcXNwkcVhw8/yA===)
``` 
using System.Threading.Tasks;

public class Program
{
    public async Task M()
    {
        await Task.Delay(1);
        GetRS().Dispose();
        return;
    }
    
    static RS GetRS() => default;
}

public ref struct RS
{
    public void Dispose() { }
}
```

The problem is that we cannot spill these types, not every usage requires spilling though.

So, I think the wording should be more specific to the situation. I would use something like: "A using statement resource of type 'RS' cannot be used in async methods or async lambda expressions."

**Comment by AlekseyTs** [comment]
Done with review pass (commit 3)

File: src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Semantic/Semantics/SpanStackSafetyTests.cs:1188]: Or, if we want to have a more general wording, something like: "Values of type 'RS' cannot be used in async methods or async lambda expressions **in this context**."

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 7]
```diff
@@ -2933,18 +2933,16 @@ private BoundExpression BindOutDeclarationArgument(DeclarationExpressionSyntax d
         }
 
         /// <summary>
-        /// Returns true if a bad special by ref local was found.
+        /// Reports an error when a bad special by-ref local was found.
         /// </summary>
-        internal static bool CheckRestrictedTypeInAsyncMethod(Symbol containingSymbol, TypeSymbol type, BindingDiagnosticBag diagnostics, SyntaxNode syntax)
+        internal static void CheckRestrictedTypeInAsyncMethod(Symbol containingSymbol, TypeSymbol type, BindingDiagnosticBag diagnostics, SyntaxNode syntax, bool reportOnTemp = false)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2938]: >reportOnTemp [](http://example.com/codeflow?start=162&length=12)

forUsingExpression?

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 4), with a parameter rename suggestion

**Pull request merged**